iree/task/task.c

```diff
@@ -0,0 +1,793 @@
+  // then the per-slice statistics will roll up into the dispatch statistics.
+  out_task->dispatch_statistics = &dispatch_task->statistics;
+  memset(&out_task->slice_statistics, 0, sizeof(out_task->slice_statistics));
+}
+
+iree_task_dispatch_slice_t* iree_task_dispatch_slice_allocate(
+    iree_task_dispatch_t* dispatch_task, const uint32_t workgroup_base[3],
+    const uint32_t workgroup_range[3], const uint32_t workgroup_count[3],
+    iree_task_pool_t* slice_task_pool) {
+  iree_task_dispatch_slice_t* slice_task = NULL;
+  iree_status_t status =
+      iree_task_pool_acquire(slice_task_pool, (iree_task_t**)&slice_task);
+  if (!iree_status_is_ok(status)) {
+    iree_status_ignore(status);
+    return NULL;
+  }
+  iree_task_dispatch_slice_initialize(dispatch_task, workgroup_base,
+                                      workgroup_range, workgroup_count,
+                                      slice_task);
+  slice_task->header.pool = slice_task_pool;
+  return slice_task;
+}
+
+iree_status_t iree_task_dispatch_slice_execute(
+    iree_task_dispatch_slice_t* task,
+    iree_task_submission_t* pending_submission) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // TODO(benvanik): coroutine support. Ideally this function can be called
+  // multiple times for the same slice, and we'll have a way to ready up the
+  // slices on the same workers (some per-worker suspended list?).
+
+  // Prepare context shared for all tiles in the slice.
+  iree_task_tile_context_t tile_context;
+  memcpy(&tile_context.workgroup_size, task->workgroup_size,
+         sizeof(tile_context.workgroup_size));
+  memcpy(&tile_context.workgroup_count, task->workgroup_count,
+         sizeof(tile_context.workgroup_count));
+  tile_context.shared_memory = task->shared_memory;
+  tile_context.statistics = &task->slice_statistics;
+
+  const uint32_t base_x = task->workgroup_base[0];
+  const uint32_t base_y = task->workgroup_base[1];
+  const uint32_t base_z = task->workgroup_base[2];
+  const uint32_t range_x = task->workgroup_range[0];
+  const uint32_t range_y = task->workgroup_range[1];
+  const uint32_t range_z = task->workgroup_range[2];
+  for (uint32_t z = base_z; z <= range_z; ++z) {
+    tile_context.workgroup_xyz[2] = z;
+    for (uint32_t y = base_y; y <= range_y; ++y) {
+      tile_context.workgroup_xyz[1] = y;
+      for (uint32_t x = base_x; x <= range_x; ++x) {
+        tile_context.workgroup_xyz[0] = x;
+        IREE_TRACE_ZONE_BEGIN_NAMED(z_tile,
+                                    "iree_task_dispatch_slice_execute_tile");
+        IREE_TRACE_ZONE_SET_COLOR(z_tile,
+                                  iree_task_tile_to_color(&tile_context));
+
+        // NOTE: these are useful for debugging but dramatically increase our
+        // cost here; only enable if needed for tracking work distribution:
+        IREE_TRACE_ZONE_APPEND_VALUE(z_tile, x);
+        IREE_TRACE_ZONE_APPEND_VALUE(z_tile, y);
+        IREE_TRACE_ZONE_APPEND_VALUE(z_tile, z);
```

Comment:
Intended to be uncommented, given the performance note?

I'm also wondering at a glance if the x/y/z are meant to be appended in the inner loop instead of up a level. Could use a bit more explanation in a comment.